src/encoding/encoding.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+use super::resource::Patch;
+use super::{DrawTag, PathEncoder, PathTag, Transform};
+
+use peniko::{kurbo::Shape, BlendMode, BrushRef, Color, ColorStop};
+
+/// Encoded data streams for a scene.
+#[derive(Default)]
+pub struct Encoding {
+    /// The path tag stream.
+    pub path_tags: Vec<PathTag>,
+    /// The path data stream.
+    pub path_data: Vec<u8>,
+    /// The draw tag stream.
+    pub draw_tags: Vec<DrawTag>,
+    /// The draw data stream.
+    pub draw_data: Vec<u8>,
+    /// Draw data patches for late bound resources.
+    pub patches: Vec<Patch>,
+    /// Color stop collection for gradients.
+    pub color_stops: Vec<ColorStop>,
+    /// The transform stream.
+    pub transforms: Vec<Transform>,
+    /// The line width stream.
+    pub linewidths: Vec<f32>,
+    /// Number of encoded paths.
+    pub n_paths: u32,
+    /// Number of encoded path segments.
+    pub n_path_segments: u32,
+    /// Number of encoded clips/layers.
+    pub n_clips: u32,
+}
+
+impl Encoding {
+    /// Creates a new encoding.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns true if the encoding is empty.
+    pub fn is_empty(&self) -> bool {
+        self.path_tags.is_empty()
```

Comment:
Partly a note to self, this will change when there are rectangles.

Reply:
I think this was originally a hack to detect empty glyphs when that would break the path encoding but I think it’s probably useful regardless.